### DISCUSSION TITLE
Planning for Scale

### DISCUSSION THREAD

**Discussion by TonyValenti** [oqtane/oqtane.framework#809] [General]
@sbwalker -
I do a lot of work with various legal technology platforms and I have a lot of knowledge about the challenges that the developers run into.

Two common pitfalls that I see large-scale app developers run into involve the use of two data types:
```int``` and ```datetime```

## ```Int``` in the Cloud
In cloud-based systems, ```int``` primary keys are often exhausted and the developers have to do extensive refactoring in order to replace int keys with long keys.  As an example, I work with attorneys and most legal tech apps preserve every email ever person in a firm sends/receives every day.  Now multiply that just by the number of people working at law firms in your region and you'll get very large numbers very quickly.  I see companies that have to deal with this challenge a lot.

Considering that we are in the age of the cloud, we should plan for large system usage.  

Are you open to a PR that replaces ```int``` with ```long``` and makes general recommendation to use ```long``` as a primary key?

## ```DateTime``` in the Cloud
Prior to working with legal tech, ```DateTime```s were the bane of my existence.  When looking at a timestamp like 2020-01-01 @ 15:30, when did it really happen?  My time zone, your time zone, the client's time zone, the server's timezone, or GMT?  Being consistent requires a great deal of vigilance.

In my industry, nearly every company I work with either has or is in the process of moving to DateTimeOffsets.  This greatly simplifies development because values are then easily converted between different timezones and you never have to ask the question "Whose timezone?" again.

Are you open to a PR that replaces ```DateTime``` with ```DateTimeOffset``` and makes a general recommendation to use ```DateTimeOffset``` when referencing dates?

**Comment by jimspillane** [comment]
I would analyze each table to determine the need for a larger primary key. For example, the _Tenant_ table might not need a _bigint_  number of tenants.

I like your proposal for modifying the DateTime.

**Comment by sbwalker** [comment]
I am sure this is obvious, but these are breaking changes and would make all existing external modules incompatible.
As a result, these changes do not align with the Oqtane philosophy for Backward Compatibility ( https://www.oqtane.org/Resources/Blog/PostId/538/oqtane-philosophy ). That being said, its useful to discuss each of these in more detail...

Int - has a max value of 2,147,483,647. For the vast majority of core entities in Oqtane the probability is extremely low of ever reaching this limit.  There are a few cases which may benefit from being a bigint - but these are exceptions rather than the norm. Note that there is nothing preventing developers from using bigints in their custom modules.

DateTime - you will notice that all DateTime fields in the core framework are populated using DateTime.UtcNow so that they are stored in a consistent format and can be reliably converted to the users time zone for evaluation or display purposes. Note that there is nothing preventing developers from using DateTimeOffset in their custom modules.

From a platform product management perspective, these changes would need to be part of a major release ( ie. a dot-0 release ) so that it clearly signals a massive breaking change. The database upgrade aspects of this change would also be quite significant and would need to be tested extensively. The scheduled release for V2.0 is Nov 10 so it does not seem realistic to consider these changes as part of that release.

**Reply by hishamco** [reply]
> The scheduled release for V2.0 is Nov 10 so it does not seem realistic to consider these changes as part of that release.

I'm agree with @sbwalker, this is not the right time for such change, probably we can consider this for the future release if it's requested by many users or it has a big value

Thanks

**Reply by PizzaConsole** [reply]
@TonyValenti  I believe this response from  Shaun can be marked as the answer to this discussion.

**Comment by louyo** [comment]
That said, it might not be a bad idea to show that (time is GMT) on the log page. All log entries are in GMT.  Where the header says Date, make it "Date (GMT)"

**Comment by louyo** [comment]
I find it odd that you would say that. But, I am no longer evaluating Oqtane.
Best of luck to all.

**Reply by TonyValenti** [reply]
@louyo - What platform did you select?  I'm digging deep into ABP.io.

**Reply by mikecasas** [reply]
@TonyValenti Is ABP like drinking from a fire hose?

**Reply by TonyValenti** [reply]
@mikecasas - ABP.io has been thought out really well and they have a lot of infrastructure in place that enterprise developers would usually end up having to develop anyways.  They have great documentation and I really like that they aren't constrained by binary compatibility handcuffs.  I think that if you're starting a new app, you'll need an extra day or so to learn the ABP.io "way" but after that, you'll be able to crank out better solutions with less code more quickly.

**Reply by jimspillane** [reply]
Hey @TonyValenti and @mikecasas, you might not have seen ABP's commercial version has a code generator _ABP Suite_ that helps create a good starting point for new entities.  I recently came across [EasyAbp](https://easyabp.io/) that has an OS code generator and modules that might help give a leg up.

**Comment by louyo** [comment]
Still searching/evaluating. We are trying to create a  central control/monitoring system, with browser access, for a variety of IoS sensing devices and controllers. Controlling and monitoring an office building. Some directly connected (eg: I2C,  DI/DO) and some via syslog server access. Latest testing effort is Blazor, running on a Linux system, also testing on one of the monitoring Raspberry Pi's.

Lou

**Reply by markdav-is** [reply]
I was able to get ROS working in blazor.  Lots of ROS messages for various sensors and such.
